File: 2021/2.jl
f = open("inputs/2.txt", "r")

lines = read(f, String)
lines = split(lines, "\r\n")

# First part
depth = 0
horizont = 0

for line in lines
    value = parse(Int, line[end])
    if line[1] == 'f'
        global horizont += value
    elseif line[1] == 'd'
        global depth += value
    elseif line[1] == 'u'
        global depth -= value
    end
end

println("First part: ", horizont * depth)


# Second part
aim = 0
depth = 0
horizont = 0

for line in lines
    value = parse(Int, line[end])
    if line[1] == 'f'
        global horizont += value
        global depth += (aim * value)
    elseif line[1] == 'd'
        global aim += value
    elseif line[1] == 'u'
        global aim -= value
    end
end
println("Second part: ", horizont * depth)

File: 2024/2.lua
function isAscending (lst)
	for ind = 1, #lst-1 do
		if lst[ind] >= lst[ind+1] then return false end
	end
	return true
end

function isDescending (lst)
	for ind = 1, #lst-1 do
		if lst[ind] <= lst[ind+1] then return false end
	end
	return true
end

function hasCorrectDifferences (lst)
	for ind = 1, #lst-1 do
		local diff = math.abs(lst[ind+1] - lst[ind])
		if diff < 1 or diff > 3 then
			return false
		end
	end

	return true
end

function isSafe (lst)
	return (isAscending(lst) or isDescending(lst)) and hasCorrectDifferences(lst)
end

function isSafeAfterRemoval (lst)
	for skipInd = 1, #lst do
		local newLst = {}
		for ind = 1, #lst do
			if ind ~= skipInd then
				table.insert(newLst, lst[ind])
			end
		end
		if isSafe(newLst) then
			return true
		end
	end
	return false
end


-- Read input file
local file = io.open("inputs/2.txt", "r");
local numSafe = 0
local numSafe_tol = 0
-- Parse the reports
for line in file:lines() do
	local report = {}
	for level in line:gmatch("%S+") do
		table.insert(report, tonumber(level))
	end
	if isSafe(report) then
		numSafe = numSafe+1
	elseif isSafeAfterRemoval(report) then
		numSafe_tol = numSafe_tol+1
	end
end

print("Number of safe reports:", numSafe)
-- Part 2

print("Number of reports within toleration:", numSafe_tol+numSafe)